modules/dcache-chimera/src/main/java/org/dcache/chimera/namespace/ChimeraCleaner.java

```diff
@@ -454,20 +457,34 @@ public void messageArrived(PoolManagerPoolUpMessage poolUpMessage)
         }
     }
 
+    /**
+     * runNotification
+     * collect messages from database and send Delete Notifications
+     */
     private void runNotification()
             throws InterruptedException
     {
         final String QUERY =
                 "SELECT ipnfsid FROM t_locationinfo_trash t1 " +
-                "WHERE itype=2 AND NOT EXISTS (SELECT 1 FROM t_locationinfo_trash t2 WHERE t2.ipnfsid=t1.ipnfsid AND t2.itype <> 2)";
-        for (String id : _db.queryForList(QUERY, String.class)) {
-            try {
-                sendDeleteNotifications(new PnfsId(id)).get();
-                _db.update("DELETE FROM t_locationinfo_trash WHERE ipnfsid=? AND itype=2", id);
-            } catch (ExecutionException e) {
-                _log.warn(e.getCause().getMessage());
+                "WHERE itype=2 AND NOT EXISTS (SELECT 1 FROM t_locationinfo_trash t2 WHERE t2.ipnfsid=t1.ipnfsid AND t2.itype <> 2) LIMIT=?";
+
+        List<String> queryList = new ArrayList<String>();
+        do {
+            queryList = _db.queryForList(QUERY, String.class, queryLimit);
+
+            if(queryList.size() == 0) {
+                break;
             }
-        }
+
+            for (String id : queryList) {
+                try {
+                    sendDeleteNotifications(new PnfsId(id)).get();
+                    _db.update("DELETE FROM t_locationinfo_trash WHERE ipnfsid=? AND itype=2", id);
+                } catch (ExecutionException e) {
+                    _log.warn(e.getCause().getMessage());
+                }
+            }
+        } while (queryList.size() >= queryLimit);
```

Comment:
I guess the real danger here is that if one of the receiving services goes down, all messages sent to that service will fail.

If the is QUERYLIMIT or more entries in the database, the while-loop will not terminate until the service comes back up again.

This has three bad effects:
  1. cleaner will send messages as fast as it can, hammering both the messaging infrastructure and the targeted services.
  2. cleaner will spam the log file will entries; again, just as fast as it can.
  3. no further cleaning of deleted pool files will take place.